src/main/java/com/hermesworld/ais/galapagos/applications/impl/UpdateApplicationAclsListener.java

```diff
@@ -265,6 +288,11 @@ public String getKafkaUserName() {
                     .filter(topic -> topic.getType() != TopicType.INTERNAL && id.equals(topic.getOwnerApplicationId()))
                     .map(topic -> topicAcls(userName, topic.getName(), WRITE_TOPIC_OPERATIONS)).forEach(result::addAll);
 
+            topicService.listTopics(environmentId).stream()
+                    .filter(topic -> topic.getType() != TopicType.INTERNAL && topic.getProducers() != null
+                            && topic.getProducers().contains(id))
+                    .map(topic -> topicAcls(userName, topic.getName(), WRITE_TOPIC_OPERATIONS)).forEach(result::addAll);
+
```

Comment:
This whole block should be combined with the one above, otherwise, listTopics() is called twice (and topics are iterated twice). Only the filter condition must be combined, rest is identical.